### ISSUE TITLE
feature request: analog of argsort in numpy

### ISSUE THREAD

**Issue by nilsbecker** [c-cube/ocaml-containers#81]
i was looking for a function `('a -> 'a -> int) -> 'a array -> int array` that gives the indices that successive array elements would have in a sorted array (sorted by the compare function given first). i.e. the ranking of the elements. something like that is useful for all kinds of rank statistics.

**Comment by c-cube** [comment]
If I understand correctly, it would map `[|"c";"d";"a";"b"|]` to `[|2;3;0;1|]`, is that it? Would you also need a function that applies the resulting `int array` to a `'a array` by permuting it, so that `Array.sort cmp a = Array.apply_perm a (Array.sorted_indices cmp a)`?

**Comment by nilsbecker** [comment]
yes, indeed it should map `[|"c";"d";"a";"b"|]` to `[|2;3;0;1|]`. let's see if i understand the permutation question. first we should always have if `asi = Array.sorted_indices cmp a` and if `a_sorted` is the sorted version of `a` then `a_sorted.(asi.(i)) = a.(i)`. 
what you are proposing is the other way around: a function so that `a_sorted.(i) = (apply_perm a asi).(i)`? does this add anything over just sorting `a` directly?  

by the way, the `Array.sort` function is always in-place, maybe that should get a functional version which would then basically do what you say directly?

on the other hand, maybe applying general permutations is generally useful also in other contexts. i guess then one has to check limits on the integer array?

**Comment by c-cube** [comment]
The permutation question is useful as a algebraic law that I can easily test, and that describes the semantics of the operation. I don't propose to sort `a` this way, it's just what it means intuitively! I'm not sure the permutation part is really useful by itself,  though, and it's really easy (just `Array.map (Array.get a) b`) so I will wait until someone explicitely asks for it. Working on the `arg_sort` function.

re Array.sort in place: I'm actually adding a `sorted` function right now, as it is indeed useful too.

**Comment by nilsbecker** [comment]
cool! thanks. i agree that the permutation is more elegantly applied just by mapping as you say. no need to inflate the api.

**Comment by nilsbecker** [comment]
ok, i just found out that i misunderstood how numpy argsort works: it's the inverse permutation of what i was describing. according to the documentation, numpy.argsort returns the indices that would sort an array. i.e. if `nasi = numpy.argsort a` then `a.(nasi.(i)) = a_sorted.(i)`. this is _not_ the ranking of elements in `a`.

notice that `asi` above and `nasi` are inverses of each other. now i'm confused as to which one is the generally more useful one. for me now it was `asi` what was needed to do a ranking. in the SO question below they describe how to get `asi` from `nasi`. and i think actually `Array.map (Array.get nasi) a = a_sorted` whereas `Array.map (Array.get asi) a_sorted = a`.

see also http://stackoverflow.com/questions/17901218/numpy-argsort-what-is-it-doing

**Comment by c-cube** [comment]
Ah, duh, that's what I was implementing (it's a bit confusing -_-). Do you want both?

**Comment by nilsbecker** [comment]
i guess both are useful - i confess i don't know why numpy chooses the one it chooses. actually, if the implementation of `asi` actually just does `nasi` twice, then one could also write into the documentation string of `nasi` how exactly to get the inverse, and leave away `asi`.

**Comment by c-cube** [comment]
Ok, I suggest, about the naming:

- `sort_indices` for the one such that `map (get a) (sort_indices a) = sorted a`
- `sort_ranking` for the one where `map (fun i -> (sorted a).(sort_ranking.(i))) = a`

**Comment by nilsbecker** [comment]
sounds good to me. for the second i guess you mean the function should be called `sort_ranking` so that `map (get (sorted a)) (sort_ranking a)) = a` ?

**Comment by c-cube** [comment]
Please give it a try ;-)

**Comment by nilsbecker** [comment]
thanks, i will; probably tomorrow though.

**Comment by nilsbecker** [comment]
one nitpick already: the sentence `such that [b.(i)] is the index of the [i]-th element in [sort cmp a]` in the documentation is missing 'of a' after 'ith element.

**Comment by nilsbecker** [comment]
hmm. opam complains:
`opam pin add bqdyn245_127:~ nbecker$ opam pin add containers https://github.com/c-cube/ocaml-containers.git` gives
```
# opam-version 1.2.2
# os           darwin
# command      ./configure --prefix /Users/nbecker/.opam/4.03.0+flambda --enable-thread --disable-bench --disable-tests --enable-bigarray --enable-advanced --enable-unix --enable-docs
# path         /Users/nbecker/.opam/4.03.0+flambda/build/containers.dev
# compiler     4.03.0+flambda
# exit-code    2
# env-file     /Users/nbecker/.opam/4.03.0+flambda/build/containers.dev/containers-45336-d8a4ae.env
# stdout-file  /Users/nbecker/.opam/4.03.0+flambda/build/containers.dev/containers-45336-d8a4ae.out
# stderr-file  /Users/nbecker/.opam/4.03.0+flambda/build/containers.dev/containers-45336-d8a4ae.err
### stdout ###
# ocamlfind ocamlopt -o setup.exe -linkpkg -package oasis.dynrun setup.ml || ocamlfind ocamlc -o setup.exe -linkpkg -package oasis.dynrun setup.ml || true
# rm -f setup.cmi setup.cmo setup.cmx setup.o
# ./setup.exe -configure --prefix /Users/nbecker/.opam/4.03.0+flambda --enable-thread --disable-bench --disable-tests --enable-bigarray --enable-advanced --enable-unix --enable-docs
### stderr ###
# configure options:
# [...]
#   --sbindir dir        System admin executables [$exec_prefix/sbin]
#   --bindir dir         User executables [$exec_prefix/bin]
#   --exec-prefix dir    Install architecture-dependent files in dir [$prefix]
#   --prefix dir         Install architecture-independent files dir [/usr/local]
#   --print-hidden Print even non-printable variable. (debug)  [false]
#   -help                Display this list of options
#   --help               Display this list of options
#
# make: *** [configure] Error 1
```

where the error file contains
```
none: unknown option '--enable-thread'.
configure options:
  --override var+val   Override any configuration variable.
  --enable-tests       Compile tests executable and library and run them [default: disabled]
  --disable-tests      Compile tests executable and library and run them [default: disabled]
  --enable-docs        Create documentations [default: enabled]
  --disable-docs       Create documentations [default: enabled]
  --enable-profile     Turn ocaml profile flag on [default: disabled]
  --disable-profile    Turn ocaml profile flag on [default: disabled]
  --enable-debug       Turn ocaml debug flag on [default: enabled]
  --disable-debug      Turn ocaml debug flag on [default: enabled]
  --destdir dir        Prepend a path when installing package
  --psdir dir          PS documentation [$docdir]
  --pdfdir dir         PDF documentation [$docdir]
  --dvidir dir         DVI documentation [$docdir]
  --htmldir dir        HTML documentation [$docdir]
  --docdir dir         Documentation root [$datarootdir/doc/$pkg_name]
  --mandir dir         Man documentation [$datarootdir/man]
  --localedir dir      Locale-dependent data [$datarootdir/locale]
  --infodir dir        Info documentation [$datarootdir/info]
  --datadir dir        Read-only architecture-independent data [$datarootdir]
  --datarootdir dir    Read-only arch-independent data root [$prefix/share]
  --libdir dir         Object code libraries [$exec_prefix/lib]
  --localstatedir dir  Modifiable single-machine data [$prefix/var]
  --sharedstatedir dir Modifiable architecture-independent data [$prefix/com]
  --sysconfdir dir     Read-only single-machine data [$prefix/etc]
  --libexecdir dir     Program executables [$exec_prefix/libexec]
  --sbindir dir        System admin executables [$exec_prefix/sbin]
  --bindir dir         User executables [$exec_prefix/bin]
  --exec-prefix dir    Install architecture-dependent files in dir [$prefix]
  --prefix dir         Install architecture-independent files dir [/usr/local]
  --print-hidden Print even non-printable variable. (debug)  [false]
  -help                Display this list of options
  --help               Display this list of options
```

**Comment by c-cube** [comment]
I don't know why, the last oasis release does that sometimes (I had the same error, too).

**Comment by nilsbecker** [comment]
tried from another computer -- same error. cannot test for the moment unfortunately...

**Comment by c-cube** [comment]
If you can reliably reproduce the issue with oasis, it will probably interest @gildor478 (you can also go on oasis' [gitter room](https://gitter.im/ocaml/oasis)).

**Comment by gildor478** [comment]
The problem is +flambda:
https://forge.ocamlcore.org/tracker/index.php?func=detail&aid=1674&group_id=54&atid=291

Upstream author of flambda has been notified and there will be a fix in oasis 0.4.8.
